### ISSUE TITLE
Make Design view dockable too

### ISSUE THREAD

**Issue by manuelma** [Spine-project/Spine-Toolbox#1297]
Every time I need to debug a model I'm tempted to ditch Spine toolbox and open a proper Python or Julia console. Toolbox doesn't seem appropriate to do heavy scripting and I think the problem is Design View is too preponderant. In my opinion, it doesn't need to be like this, we don't need to see the beautiful workflow constantly. It can be useful to make a console the bigger widget from time to time.

So if you don't mind, I will go ahead and make Design View dockable as the rest of the dock widgets.

**Comment by DillonJ** [comment]
I wonder... most of us who are developing models have some sort of IDE, editor, workspace manager... e.g. VS Code, Atom or others. Then you have some competition regarding who hosts the scripting interface - Toolbox or your IDE. I personally find running from VS Code and previously Atom to be a more "comfortable" experience... and many of the reasons for this are perhaps superficial... the VS code environment just feels nicer.

Do you think there is a way we can link these two ways for working in a more sympathetic way. What you have suggested above sounds like a first step. Do we need a file browser? But perhaps that is overlapping too much with IDE functions. 

I suppose I'm wondering if we want to brainstorm here. Might even be a longer term objective? What about a VS code extension for toolbox?

**Comment by DillonJ** [comment]
Responding more directly to the specific issue - I think it's  a really good idea!

**Comment by manuelma** [comment]
At the moment if you double click on a Tool item in design view, the Tool spec editor opens up and you get an 'okish' text browser with your code. It also has syntax highlighting for Julia and Python since recently. I think the tool spec editor could be make more like an IDE, with the source files list on the left. You could select those files and see the code in the mentioned text browser.

**Comment by soininen** [comment]
I think we should try to keep code editing features in Toolbox at bare minimum. There are so many great IDEs out there it just doesn't make sense to try to duplicate that functionality. Especially not in Tool specification editor.

With regards to interoperability with IDEs, we have the 'headless' mode which allows one to execute a project directly from command line. If your IDE supports it, you could use headless to run your workflow directly from your IDE.

**Comment by manuelma** [comment]
Those are good points. The problem comes when you're trying to do some modelling. Switching between ide and toolbox is not very comfortable. Typically you'd chose "Open main program" from Toolbox, and it will open the file in a tab in your ide. Your ide already has multiple tabs from other things you're doing, you start to get lost. I think it would be better if Tool spec editor allowed you to see all your program files, edit and save them if necessary. Like a small project manager with as many features as possible, not trying to replicate of course the full functionality of an ide, but some of the very basic ones.

**Comment by manuelma** [comment]
Using the headless mode from your ide is a good idea, but only when you're working on your code. Not when you also want to modify importer/exporter specs, view/edit the contents of the database, add project items and modify your workflow at the same time.

**Comment by soininen** [comment]
>  Like a small project manager with as many features as possible, not trying to replicate of course the full functionality of an ide, but some of the very basic ones.

Basic editing capabilities are OK. There just needs to be one conservative developer to keep feature creep at bay ;)

Maybe the editor could be a separate widget so we could have it as a dockable window in Toolbox main window as well?

**Comment by PekkaSavolainen** [comment]
I actually tried this a long time ago but I'm seeing the same problems here that made me drop the whole thing. For example, when you maximize the screen, Event log and Python consoles takes half the screen and they cannot be made smaller anymore.

![issue_1297_dockabledesignview](https://user-images.githubusercontent.com/43600314/111284297-accfd100-8648-11eb-93d1-39a975003502.png)

**Comment by PekkaSavolainen** [comment]
Basically the problem is that the dock widgets need a central widget that acts as an anchor for them. If the problem you are trying to solve here is that you want to make something else the 'central widget' (like the mini IDE), I propose that you just make a tabwidget as the central widget and put the the Design View there. Then you can add other tabs for the IDE windows.

**Comment by manuelma** [comment]
Which commit are you on?

**Comment by manuelma** [comment]
>  I propose that you just make a tabwidget as the central widget and put the the Design View there. Then you can add other tabs for the IDE windows.

I'm not sure I understand. I want to exchange Julia console and Design view in your screenshot above. I don't know how to achieve that with the tab widget approach. Also, I don't experience the 'problems' you do, and I wonder if it's because you didn't check out the commit where we enable dockwidget-nesting. This is how it looks in my system:

![Screenshot from 2021-03-16 10-42-03](https://user-images.githubusercontent.com/11472756/111288312-6e232200-861a-11eb-9966-d39ed1eea967.png)

**Comment by manuelma** [comment]
And to clarify, this is what I'm after (and can do thanks to this change): A big console where I can comfortably debug, while still being able to see the workflow and the item:

![Screenshot from 2021-03-16 10-50-22](https://user-images.githubusercontent.com/11472756/111289544-a6773000-861b-11eb-81f2-4e93a70251e7.png)

**Comment by PekkaSavolainen** [comment]
commit 32ef309

**Comment by manuelma** [comment]
Weird. I tried it on windows and it also looks good:

![Screenshot from 2021-03-16 11-05-23](https://user-images.githubusercontent.com/11472756/111291549-af690100-861d-11eb-920d-72662ec0f081.png)

**Comment by PekkaSavolainen** [comment]
You could do this (in commit c8713598)

![issue_1297_proposal](https://user-images.githubusercontent.com/43600314/111291543-f839ad80-864f-11eb-9df9-bba9f1eccb1e.png)

**Comment by manuelma** [comment]
I don't think Qt would provide `QMainWindow.takeCentralWidget()` if it wasn't intended to work correctly?

**Comment by manuelma** [comment]
> You could do this (in commit c871359)

**Comment by PekkaSavolainen** [comment]
Move the dockwidgets around, resize the dock widgets by grabbing the separator, resize the window (maximize and go back), now try to make the window look like in comment above where you said "Weird. I tried it on windows and it also looks good (commit 32ef309):"

**Comment by manuelma** [comment]
Can you be more specific? I don't see any issues doing any dock reconfiguration on windows.

Also, are you sure the issue you describe is related to `takeCentralWidget()`? We use it in all our other main windows (Spine DB editor, and every specification editor), and it works just fine.

**Comment by PekkaSavolainen** [comment]
hmm, looks like the problems I had, were related to some remnant dockwidget places that were stored in QSettings, after manually removing the settings I can now see the changes better. Now, the separators work as before and it is quite flexible, Just needs a bit of manual labor from our users to set it up like they want it but that's fine. I can live with this. Good work.

**Comment by soininen** [comment]
> some remnant dockwidget places that were stored in QSettings,

Is there a danger users have similar settings out there and will run into similar problems? Can we somehow avoid that?

**Comment by manuelma** [comment]
> Can we somehow avoid that?

I think we can, by skipping `restore_ui()` only once after users checked this commit.

**Comment by PekkaSavolainen** [comment]
Not sure if my registry is in anyway comparable to a regular users registry though :)

**Comment by manuelma** [comment]
If that registry could talk...

Ok, I'll keep moving widgets around to try and see the issues @PekkaSavolainen describes. For me it's relatively important to make this work...

**Comment by jkiviluo** [comment]
Nice improvement. I like the idea of seeing all project files too. I suppose it's a bit bigger piece of work though.

**Comment by jkiviluo** [comment]
@manuelma Is there still an issue here?

**Comment by manuelma** [comment]
There is a problem with resizing widgets still, but we have another issue for that.
